### PR TITLE
LVM-424: Add new fields to student database

### PR DESCRIPTION
**Pull request by vcantu**
Takes care of the tickets:
[LVM-424: Add new fields to student databse](https://cs5500-jira.ccs.neu.edu/browse/LVM-424)
[LVM-308: Write rspec for student model](https://cs5500-jira.ccs.neu.edu/browse/LVM-308)

Adds following fields to student database:
+ status
+ status_date_of_change
+ status_changed_by
+ last_name_id
+ preferred_contact
+ immigrant_status
+ education
+ services_requested
+ additional_services_requested
+ criminal_conviction
+ release_on_file
+ release_signed_date
+ cdbg_required
+ cdbg_us_citizen
+ cdbg_legal_resident
 + cdbg_female_head_of_household
+ cdbg_household_size
+ cdbg_household_income
- tutor_preference
+ age_preference
+ meet_at_local_library
+ where_Can_meet
+ transportation
+ other_preferences

Also adds model validations and rspec for student model for new fields

### CODE REVIEW

File: app/models/student.rb [hunk 1]
```diff
@@ -29,8 +36,42 @@ class Student < ApplicationRecord
...
+  # ======================
+  def all_tags=(names)
+    self.tags = names.reject(&:empty?).uniq.map do |name|
+      Tag.where(name: name.strip).first_or_create!
+    end
+  end
+
+  def all_tags
+    tags.map(&:name)
+  end
```
Comment (zack) [inline app/models/student.rb:84]: Students get tags? :o

**Comment by zack** [comment]
looks like yah got some build failures.

`ActiveRecord::StatementInvalid: SQLite3::ConstraintException: NOT NULL constraint failed: students.gender: INSERT INTO "students" ("created_at", "updated_at", "first_name", "last_name", "email", "cell_phone", "home_phone", "work_phone") VALUES (?, ?, ?, ?, ?, ?, ?, ?)`

File: app/models/student.rb (hunk 1 above)
Comment (vcantu) [reply app/models/student.rb:84]: I thought tags were also for students, I can removed them

Comment (nabeel-) [reply app/models/student.rb:84]: Yes, they should have tags as well as comments

File: app/models/student.rb [hunk 2]
```diff
@@ -7,15 +13,16 @@ class Student < ApplicationRecord
...
   has_many :tags, through: :taggings
 
   validates :first_name, presence: true
-  # validates :last_name, presence: true
+  validates :last_name, presence: true
   # validates :dob, presence: true
-  # validates :gender, presence: true
-  # validates :smartt_id, presence: true
+  validates :gender, presence: true
+  validates :smartt_id, presence: true
```
Comment (nabeel-) [inline app/models/student.rb]: It should allow for a blank `smartt_id` for the same reason as a tutor: they are only given when entering the data into smartt and not necessarily known at creation time in our portal.

File: app/controllers/students_controller.rb [hunk 3]
```diff
@@ -60,7 +60,7 @@ def set_tutor
 
   def student_params
     params.require(:student).permit(
-      :first_name
+      :first_name, :last_name, :gender
```
Comment (zack) [inline app/controllers/students_controller.rb]: This should have every parameter we allow the user to type into the form.

File: app/models/student.rb [hunk 4]
```diff
@@ -7,14 +13,15 @@ class Student < ApplicationRecord
   has_many :assessments
 
   has_many :affiliates
+  has_many :student_comments
```
Comment (zack) [inline app/models/student.rb]: This isn't a model?

File: app/models/student.rb [hunk 5]
```diff
@@ -1,4 +1,10 @@
 class Student < ApplicationRecord
+  VALID_EMAIL_REGEX  = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  VALID_PHONE_REGEX  = /\A\([0-9]{3}\) [0-9]{3}-[0-9]{4}\z/
+  VALID_ZIP_REGEX    = /\A[0-9]{5}\z/
+  VALID_SMARTT_REGEX = /\A[0-9]{4}-[0-9]{6}\z/
+  LAST_NAME_ID_REGEX = /\A[0-9]{4,5}\z/
```
Comment (zack) [inline app/models/student.rb:6]: I know you didn't make this up, but I think it's weird that some of these start with `VALID`, as if we ever wanted to use `INVALID` regexes.

File: db/migrate/20170328215033_update_new_student_fields.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    add_column :students, :cdbg_household_size, :integer
+    add_column :students, :cdbg_household_income, :integer
+    
+    # smartt required fields
+    change_column_null :students, :first_name, false
+    change_column_null :students, :last_name, false
+    change_column_null :students, :gender, false
+
+    # add student comments
+    create_table :student_comments do |t|
```
Comment (zack) [inline db/migrate/20170328215033_update_new_student_fields.rb]: Creating a whole new model is probably a good thing to do in its own pr. You're also missing the actual model, its controller, its validations, etc. I would suggest taking this out and doing it in another ticket. There's a lot of work for this.

**Comment by zack** [comment]
I'm still getting a few failures when I `rake db:reseed` with this code.

File: app/models/student.rb (hunk 5 above)
Comment (vcantu) [reply app/models/student.rb:6]: I can change the names but I think  `VALID` short for `VALIDATES`

Comment (xingyif) [reply app/models/student.rb:6]: I think the name is good enough. if you change one, you need to sync all other files.

File: spec/views/student/edit_spec.rb [hunk 7]
```diff
@@ -4,7 +4,8 @@
   describe 'populate page' do
     before do
       @student = Student.create(first_name: 'Joe',
```
Comment (zack) [inline spec/views/student/edit_spec.rb:6]: we should be using `create(:student {...})` here.

File: spec/views/student/index_spec.rb [hunk 8]
```diff
@@ -3,8 +3,10 @@
 RSpec.describe 'students/index.html.erb', type: :view do
   describe 'populate page and links' do
     before do
-      @students = [create(:student, first_name: 'Joe', last_name: 'Lally'),
-                   create(:student, first_name: 'Tim', last_name: 'Bradley')]
+      @students = [create(:student, first_name: 'Joe', last_name: 'Lally',
```
Comment (zack) [inline spec/views/student/index_spec.rb:6]: we should be using `create(:student {...})` here.

File: spec/views/student/new_spec.rb [hunk 9]
```diff
@@ -4,7 +4,8 @@
   describe 'populate page' do
     before do
       @student = Student.create(first_name: 'Joe',
```
Comment (zack) [inline spec/views/student/new_spec.rb:6]: we should be using `create(:student {...})` here.

File: spec/views/student/show_spec.rb [hunk 10]
```diff
@@ -6,8 +6,10 @@
       @coordinator = create(:coordinator)
       @tutors = [Tutor.create(first_name: 'Bob', last_name: 'Smith'),
                  Tutor.create(first_name: 'Tom', last_name: 'Brady')]
-      @students = [Student.create(first_name: 'Mike', last_name: 'White'),
-                   Student.create(first_name: 'Alex', last_name: 'Fallah')]
+      @students = [Student.create(first_name: 'Mike', last_name: 'White',
```
Comment (zack) [inline spec/views/student/show_spec.rb:9]: we should be using `create(:student {...})` here.

File: spec/views/student/show_spec.rb [hunk 11]
```diff
@@ -6,8 +6,10 @@
       @coordinator = create(:coordinator)
       @tutors = [Tutor.create(first_name: 'Bob', last_name: 'Smith'),
```
Comment (zack) [inline spec/views/student/show_spec.rb:7]: we should be using `create(:tutor {...})` here.

File: spec/models/student_spec.rb [hunk 12]
```diff
@@ -32,14 +32,214 @@
...
+        expect(student.age_preference_array).to eq [1, 2, 0, 0, 16]
+      end
+    end
+  end
+
+  describe '#all_tags' do
+    before do
+      @student = create(:student)
+      @tag = Tag.create(name: 'Donor')
+      Tagging.create(student_id: @student.id, tag_id: @tag.id)
```
Comment (zack) [inline spec/models/student_spec.rb:189]: We should be using `create(:tagging, {...})` here.

**Review by zack — APPROVED** [review]
nonblocking comments
